extra/modules/fiftyone-devicedetection/src/main/java/org/prebid/server/hooks/modules/fiftyone/devicedetection/v1/core/DeviceEnricher.java

```diff
@@ -0,0 +1,264 @@
+import fiftyone.pipeline.engines.data.AspectPropertyValue;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.collections4.MapUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.boundary.CollectedEvidence;
+import org.prebid.server.proto.openrtb.ext.request.ExtDevice;
+
+import jakarta.annotation.Nonnull;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+public class DeviceEnricher {
+    public static final String EXT_DEVICE_ID_KEY = "fiftyonedegrees_deviceId";
+
+    private static final Map<String, Integer> DEVICE_FIELD_MAPPING = Map.ofEntries(
+            Map.entry("Phone", OrtbDeviceType.PHONE.ordinal()),
+            Map.entry("Console", OrtbDeviceType.SET_TOP_BOX.ordinal()),
+            Map.entry("Desktop", OrtbDeviceType.PERSONAL_COMPUTER.ordinal()),
+            Map.entry("EReader", OrtbDeviceType.PERSONAL_COMPUTER.ordinal()),
+            Map.entry("IoT", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
+            Map.entry("Kiosk", OrtbDeviceType.OOH_DEVICE.ordinal()),
+            Map.entry("MediaHub", OrtbDeviceType.SET_TOP_BOX.ordinal()),
+            Map.entry("Mobile", OrtbDeviceType.MOBILE_TABLET.ordinal()),
+            Map.entry("Router", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
+            Map.entry("SmallScreen", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
+            Map.entry("SmartPhone", OrtbDeviceType.MOBILE_TABLET.ordinal()),
+            Map.entry("SmartSpeaker", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
+            Map.entry("SmartWatch", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
+            Map.entry("Tablet", OrtbDeviceType.TABLET.ordinal()),
+            Map.entry("Tv", OrtbDeviceType.CONNECTED_TV.ordinal()),
+            Map.entry("Vehicle Display", OrtbDeviceType.PERSONAL_COMPUTER.ordinal()));
+
+    private final Pipeline pipeline;
+
+    public DeviceEnricher(@Nonnull Pipeline pipeline) {
+        this.pipeline = Objects.requireNonNull(pipeline);
+    }
+
+    public EnrichmentResult populateDeviceInfo(Device device, CollectedEvidence collectedEvidence) {
+        try (FlowData data = pipeline.createFlowData()) {
+            data.addEvidence(pickRelevantFrom(collectedEvidence));
+            data.process();
+            final DeviceData deviceData = data.get(DeviceData.class);
+            if (deviceData == null) {
+                return null;
+            }
+            final Device properDevice = Optional.ofNullable(device).orElseGet(() -> Device.builder().build());
+            return patchDevice(properDevice, deviceData);
+        } catch (Exception e) {
+            return EnrichmentResult.builder().processingException(e).build();
```

Comment:
the exception is handled and returned, but never used further, do you really need it?

Reply:
This is a great question.  Maybe the design decision here was not the best.  Here is Max's thinking on the question:

:question:> Do we actually need to handle the exception this way?

What is the exact contract on exception handling within `payloadUpdate` ?

- Should the module wrap the pipeline exception into `RuntimeException` and let it bubble up and be caught by [[GroupResult.applyPayloadUpdate](https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/hooks/execution/GroupResult.java#L109)] within module's host (server app) ?
  - Will the rest of the auction continue as though the module didn't interfere at all?
- Should the `payloadUpdate` handler amend failures or tags to the already-returned `InvocationResult` ?
- Is the module responsible for logging errors inside `payloadUpdate`?
  - Should they be aggregated by some criteria (e.g. publisher account) ?

The exception was saved in the `EnrichmentResult` instance to be handled later as this didn't seem to be critical before.   Maybe we are just overcomplicating stuff and it should be just rethrown, or absorbed completely...  

Please share your thoughts, thank you.

Reply:
My initial point here is based on the fact that the exception you catch is never used. So from the code perspective we usually get rid of the stuff that are never used (I believe that's obvious). So there are two ways: 1) catch and do nothing and 2) catch and do something (here I'm mainly describing the exceptions that disrupts the module in general, so the module logic can not be applied at all). It's never worth (in both cases) to throw the exception from the module without handling it, since it's worth not to update the payload at all and return a debug/error message and no_update action status instead in the exception case.

Do nothing is basically what happens in your case, looks like you ignore the exception at all, if it's the desired behavior I can only suggest remove the `processedException` field from the `EnrichmentResult` and, as an idea, create an EMPTY EnrichmentResult to return in this case.

Do something is really depends on your desire. For example, you can log the exception for the observability purpose OR you can add a debug or error message to the invocation result and notify the client on the response.

Reply:
Thanks for the provided explanation.  

:heavy_check_mark: Removed this code.
:arrow_right: Added `throws Exception` to `DeviceEnricher.populateDeviceInfo` and `RawAuctionRequestHook.enrichDevice` method signatures.
:arrow_right: Moved `catch` clause from `try-with-resource` in `DeviceEnricher.populateDeviceInfo` to `RawAuctionRequestHook.updatePayload`.